src/Data/Tree/AVL/Indexed/Relation/Unary/Any/Properties.agda

```diff
@@ -0,0 +1,180 @@
+joinʳ⁺-here⁺ k₂ t₁ (1# , t₃)                       ∼0  p = here p
+joinʳ⁺-here⁺ k₂ t₁ (1# , t₃)                       ∼-  p = here p
+joinʳ⁺-here⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼+)         ∼+  p = left (here p)
+joinʳ⁺-here⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼0)         ∼+  p = left (here p)
+joinʳ⁺-here⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  p = left (here p)
+
+joinʳ⁺-left⁺ : ∀ {l u hˡ hʳ h} →
+              (kv : K& V) →
+              (l : Tree V l [ kv .key ] hˡ) →
+              (r : ∃ λ i → Tree V [ kv .key ] u (i ⊕ hʳ)) →
+              (bal : hˡ ∼ hʳ ⊔ h) →
+              Any P l → Any P (proj₂ (joinʳ⁺ kv l r bal))
+joinʳ⁺-left⁺ k₂ t₁ (0# , t₃)                       bal p = left p
+joinʳ⁺-left⁺ k₂ t₁ (1# , t₃)                       ∼0  p = left p
+joinʳ⁺-left⁺ k₂ t₁ (1# , t₃)                       ∼-  p = left p
+joinʳ⁺-left⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼+)         ∼+  p = left (left p)
+joinʳ⁺-left⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼0)         ∼+  p = left (left p)
+joinʳ⁺-left⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  p = left (left p)
+
+joinʳ⁺-right⁺ : ∀ {l u hˡ hʳ h} →
+                (kv : K& V) →
+                (l : Tree V l [ kv .key ] hˡ) →
+                (r : ∃ λ i → Tree V [ kv .key ] u (i ⊕ hʳ)) →
+                (bal : hˡ ∼ hʳ ⊔ h) →
+                Any P (proj₂ r) → Any P (proj₂ (joinʳ⁺ kv l r bal))
+joinʳ⁺-right⁺ k₂ t₁ (0# , t₃)                       bal p                = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , t₃)                       ∼0  p                = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , t₃)                       ∼-  p                = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼+)         ∼+  (here p)         = here p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼+)         ∼+  (left p)         = left (right p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼+)         ∼+  (right p)        = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼0)         ∼+  (here p)         = here p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼0)         ∼+  (left p)         = left (right p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼0)         ∼+  (right p)        = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (here p)         = right (here p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (left (here p))  = here p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (left (left p))  = left (right p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (left (right p)) = right (left p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (right p)        = right (right p)
+
+------------------------------------------------------------------------
+-- insert
+
+module _ {V : Value v} where
+
+  private Val = Value.family V
+
+  Any-insertWith :
+    ∀ {l u h} (k : Key) →
+    (f : Maybe (Val k) → Val k) →
+    (Pf : ∀ {k′} → (eq : k ≈ k′) (mk′ : Maybe (Val k′)) →
+          maybe′ (λ v → P (k′ , Value.respects V eq (f (just (Value.respects V (sym eq) v)))))
+                 (P (k , f nothing)) mk′)
```

Comment:
Ugh, this type is pretty unpleasant. Feels like it might be a symptom of some suboptimal or missing definition? At a minimum you can pull out `Value.respects V` into the private block above...

Reply:
Hmm maybe I've found the problem. The proof doesn't actually require the `eq : k ≈ k′` argument to go through (`refl` is always provided`) right? So it's really combining two proofs in one. Firstly that `insertWith` can introduce `Any` and secondly that `Any` respects the underlying equality. The second proof should be split out and proved separately right?

Reply:
The problem is that you don't want to end up with a proof of
`P (k , Value.respects V (sym refl) (f nothing))` when you
need a proof of `P (k , f nothing)`.

I did not want to add an extra coherence constraint stating that
`Value.respects V (sym refl)` preserves proofs.

I should perhaps just split this `Any-insertWith` into two distinct
lemmas depending on whether the key is already present in the tree
or not.

Reply:
> I should perhaps just split this Any-insertWith into two distinct
lemmas depending on whether the key is already present in the tree
or not.

That sounds like a good plan!